src/lib.rs

```diff
@@ -531,14 +658,61 @@ mod tests {
     }
 
     #[test]
+    fn memchr_iter() {
+        let haystack = b"aaaabaaaab";
+        let mut memchr_iter = Memchr::new(b'b', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next();
+        let third = memchr_iter.next();
+        assert_eq!(Some(5), first);
+        assert_eq!(Some(10), second);
+        assert_eq!(None, third);
+    }
+
+    #[test]
+    fn memchr2_iter() {
+        let haystack = b"ab";
+        let mut memchr_iter = Memchr2::new(b'a', b'b', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next();
+        let third = memchr_iter.next();
+        assert_eq!(Some(1), first);
+        assert_eq!(Some(2), second);
+        assert_eq!(None, third);
+    }
+
+    #[test]
+    fn memchr3_iter() {
+        let haystack = b"abc";
+        let mut memchr_iter = Memchr3::new(b'a', b'b', b'c', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next();
+        let third = memchr_iter.next();
+        let fourth = memchr_iter.next();
+        assert_eq!(Some(1), first);
+        assert_eq!(Some(2), second);
+        assert_eq!(Some(3), third);
+        assert_eq!(None, fourth);
+    }
+
+    #[test]
+    fn memchr_reverse_iter() {
+        let haystack = b"aaaabaaaab";
+        let mut memchr_iter = Memchr::new(b'b', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next_back();
+        let third = memchr_iter.next();
+
+        assert_eq!(Some(5), first);
+        assert_eq!(Some(10), second);
+        assert_eq!(None, third);
```

Comment:
I believe this looks right. Could you add a third `b` to `haystack` and run another `next_back` call?